ledger/internal/eval_blackbox_test.go

```diff
@@ -962,6 +962,57 @@ func benchConsensusRange(b *testing.B, start, stop int, bench func(t *testing.B,
 	}
 }
 
+// TestHeaderAccess tests FirstValidTime and `block` which can access previous
+// block headers.
+func TestHeaderAccess(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	genBalances, addrs, _ := ledgertesting.NewTestGenesis()
+	// Added in v33
+	testConsensusRange(t, 33, 0, func(t *testing.T, ver int) {
+		cv := consensusByNumber[ver]
+		dl := NewDoubleLedger(t, genBalances, cv)
+		defer dl.Close()
+
+		fvt := txntest.Txn{
+			Type:            "appl",
+			Sender:          addrs[0],
+			FirstValid:      0,
+			ApprovalProgram: "txn FirstValidTime",
+		}
+		dl.txn(&fvt, "round 0 is not available")
+
+		// advance current to 2
+		pay := txntest.Txn{Type: "pay", Sender: addrs[0], Receiver: addrs[0]}
+		dl.fullBlock(&pay)
+
+		fvt.FirstValid = 1
+		dl.txn(&fvt, "round 0 is not available")
+
+		fvt.FirstValid = 2
+		dl.txn(&fvt) // current becomes 3
+
+		// Advance current round far enough to test access MaxTxnLife ago
+		for i := 0; i < int(config.Consensus[cv].MaxTxnLife); i++ {
+			dl.fullBlock()
+		}
+		// current should be 1003. Confirm.
+		fvt.FirstValid = 1004
+		fvt.LastValid = 1010
+		dl.txn(&fvt, "txn dead")
```

Comment:
I assume we'd only get the txn dead message if 1003 was available?

Reply:
Here `txn dead` refers to the fact that fv is larger than Latest().  This was my dumb way to assert the current round, because I didn't think to do dl.Latest(), as Pavel suggests below.